src/libs/actions/Policy.js

```diff
@@ -1458,6 +1463,403 @@ function buildOptimisticPolicyRecentlyUsedCategories(policyID, category) {
     return lodashUnion([category], policyRecentlyUsedCategories);
 }
 
+/**
+ * This flow is used for bottom up flow converting IOU report to an expense report. When user takes this action,
+ * we create a Collect type workspace when the person taking the action becomes an owner and an admin, while we
+ * add a new member to the workspace as an employee and convert the IOU report passed as a param into an expense report.
+ *
+ * @param {Object} iouReport
+ * @returns {String} policyID of the workspace we have created
+ */
+function createWorkspaceFromIOUPayment(iouReport) {
+    // This flow only works for IOU reports
+    if (!ReportUtils.isIOUReport(iouReport)) {
+        return;
+    }
+
+    // Generate new variables for the policy
+    const policyID = generatePolicyID();
+    const workspaceName = generateDefaultWorkspaceName(sessionEmail);
+    const employeeAccountID = iouReport.ownerAccountID;
+    const employeeEmail = iouReport.ownerEmail;
+    const {customUnits, customUnitID, customUnitRateID} = buildOptimisticCustomUnits();
+    const oldPersonalPolicyID = iouReport.policyID;
+    const iouReportID = iouReport.reportID;
+
+    const {
+        announceChatReportID,
+        announceChatData,
+        announceReportActionData,
+        announceCreatedReportActionID,
+        adminsChatReportID,
+        adminsChatData,
+        adminsReportActionData,
+        adminsCreatedReportActionID,
+        expenseChatReportID: workspaceChatReportID,
+        expenseChatData: workspaceChatData,
+        expenseReportActionData: workspaceChatReportActionData,
+        expenseCreatedReportActionID: workspaceChatCreatedReportActionID,
+    } = ReportUtils.buildOptimisticWorkspaceChats(policyID, workspaceName);
+
+    // Create the workspace chat for the employee whose IOU is being paid
+    const employeeWorkspaceChat = createPolicyExpenseChats(policyID, {[employeeEmail]: employeeAccountID}, true);
+    const newWorkspace = {
+        id: policyID,
+
+        // We are creating a collect policy in this case
+        type: CONST.POLICY.TYPE.TEAM,
+        name: workspaceName,
+        role: CONST.POLICY.ROLE.ADMIN,
+        owner: sessionEmail,
+        isPolicyExpenseChatEnabled: true,
+
+        // Setting the currency to USD as we can only add the VBBA for this policy currency right now
+        outputCurrency: CONST.CURRENCY.USD,
+        pendingAction: CONST.RED_BRICK_ROAD_PENDING_ACTION.ADD,
+        customUnits,
+    };
+
+    const optimisticData = [
+        {
+            onyxMethod: Onyx.METHOD.MERGE,
```

Comment:
Thanks!

Reply:
😢 This is multiple times I've seen this used and I think it is a hack and that we should really try to find some other real solution to the problem.

Reply:
I did raise that concern in the linked issue, however, no other viable solution has been found in couple of weeks. Then given we already use the `SET` for classic policy creation (so not part of the bottom up flow), I agree to proceed with using `SET`.

Reply:
Would you be OK with creating a GH issue with a solid problem statement and start getting some proposals for fixing it? I'd like to not just let this one fade away with others continuing to copy the same hack. I have some ideas like:

- If there was just a `merge()` then the data should be immediately (and synchronously) available in the Onyx cache. Let's find a way to use it faster
- Is the problem because subscribers aren't notified until the next-tick? If so, maybe we can add an option to Onyx to bypass for some scenarios (since it was only added as an optimization)

Reply:
I have this saved, but I still did not get around to do this focusing on Ideal nav and waves in general. Noting previous comment from [another PR](https://github.com/Expensify/App/pull/30582#discussion_r1395850276) for future.